tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,291 @@
+    // TODO: special visitors for various formats that honeycomb.io supports
+    fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+        let s = format!("{:?}", value);
+        self.0.insert(field.name(), json!(s));
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: Data) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, Data>> {
+        self.spans.get(id_to_idx(id))
+    }
+
+    /// Decrements the reference count of the span with the given `id`, and
+    /// removes the span if it is zero.
+    ///
+    /// The allocated span slot will be reused when a new span is created.
+    fn drop_span(&self, id: Id) -> bool {
+        if !self.drop_ref(&id) {
+            return false;
+        }
+
+        // Synchronize only if we are actually removing the span (stolen
+        // from std::Arc);
+        fence(Ordering::Acquire);
+        self.spans.remove(id_to_idx(&id))
+    }
+
+    fn drop_ref(&self, id: &Id) -> bool {
+        let idx = id_to_idx(id);
+        let span = self.spans.get(idx).expect("Span not found");
+
+        let refs = span.ref_count.fetch_sub(1, Ordering::Release);
+        assert!(
+            if std::thread::panicking() {
+                // don't cause a double panic, even if the ref-count is wrong...
+                true
+            } else {
+                refs != std::usize::MAX
+            },
+            "reference count overflow!"
+        );
+        refs == 1
+    }
+}
+
+thread_local! {
+    static CURRENT_SPANS: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let mut values = HashMap::new();
+        let mut visitor = RegistryVisitor(&mut values);
+        attrs.record(&mut visitor);
+        let parent = attrs.parent().map(|id| self.clone_span(id));
+        let s = Data {
+            metadata: attrs.metadata(),
+            parent,
+            children: vec![],
+            ref_count: AtomicUsize::new(1),
+            values: Mutex::new(values),
+        };
+        let id = self.insert(s).expect("Unable to allocate another span");
+        idx_to_id(id)
+    }
+
+    #[inline]
+    fn record(&self, _: &span::Id, _: &span::Record<'_>) {
+        // TODO(david)—implement this.
+    }
+
+    fn record_follows_from(&self, _: &span::Id, _: &span::Id) {
+        // TODO(david)—implement this.
+    }
+
+    fn event(&self, event: &Event<'_>) {
+        let id = match event.parent() {
+            Some(id) => Some(id.clone()),
+            None => {
```

Comment:
TIOLI/Style: could collapse the `if` into the match:
```rust
let id = match event.parent() {
    Some(id) => Some(id.clone()),
    None if event.is_contextual() => 
        CURRENT_SPANS.with(|spans| spans.borrow().current().map(|id| id.clone())),
    None => None,
}
```

Reply:
oh yeah, +1 to flattening.